### PR TITLE
feat[next]: iterator.embedded with new Field implementation

### PR DESCRIPTION
**Pull request by havogt**
Makes iterator.embedded work with the new `common.Field` implementation.

- Temporarily `iterator.embedded.np_as_located_field` returns `common.Field`. In a next step all users should switch to a different allocation function.
- `iterator.embedded` wraps the `common.Field` into a `LocatedField` wrapper to do some translation between `iterator.embedded` `field_getitem`/`field_setitem`. This layer should eventually be removed.
- Adds a minimal (= only with requirements for `iterator.embedded`) implementation of `IndexField` and `ConstantField` to `iterator.embedded`. These can be generalized to proper `Field`s suitable for fieldview embedded.

### CODE REVIEW

File: tests/next_tests/integration_tests/feature_tests/iterator_tests/test_builtins.py [hunk 1]
```diff
@@ -363,4 +363,4 @@ def sten_cast(value):
     )
 
     if validate:
-        assert out[0] == expected_value
+        assert out.ndarray[0] == expected_value
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/iterator_tests/test_builtins.py:366]: This is an example where I think it makes sense to extract the low-level buffer.

File: tests/next_tests/integration_tests/feature_tests/iterator_tests/test_horizontal_indirection.py [hunk 2]
```diff
@@ -76,7 +76,7 @@ def test_simple_indirection(program_processor):
 
     ref = np.zeros(shape)
     for i in range(shape[0]):
-        ref[i] = inp[i + 1 - 1] if cond[i] < 0 else inp[i + 1 + 1]
+        ref[i] = inp.ndarray[i + 1 - 1] if cond.ndarray[i] < 0 else inp.ndarray[i + 1 + 1]
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/iterator_tests/test_horizontal_indirection.py:79]:
```suggestion
        ref[i] = inp[{IDim: i + 1 - 1}] if cond[{IDim: i}] < 0 else inp[{IDim: i + 1 + 1}]
```
I guess we want this.

File: tests/next_tests/integration_tests/feature_tests/iterator_tests/test_implicit_fencil.py [hunk 3]
```diff
@@ -75,7 +75,7 @@ def fun(inp0, inp1):
     run_processor(fun[dom], program_processor, inp0, inp1, out=out, offset_provider={})
 
     if validate:
-        assert np.allclose(inp0.array() + inp1.array(), out)
+        assert np.allclose(np.asarray(inp0) + np.asarray(inp1), out)
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/iterator_tests/test_implicit_fencil.py:78]:
```suggestion
        assert np.allclose(inp0 + inp1, out)
```
probably this should already work

File: tests/next_tests/integration_tests/feature_tests/iterator_tests/test_scan.py [hunk 4]
```diff
@@ -61,4 +61,4 @@ def wrapped(inp):
     )
 
     if validate:
-        assert np.allclose(out[:, :-1], reference)
+        assert np.allclose(out.ndarray[:, :-1], reference)
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/iterator_tests/test_scan.py:64]: Don't know what makes sense here.

File: tests/next_tests/integration_tests/feature_tests/iterator_tests/test_trivial.py [hunk 5]
```diff
@@ -66,7 +66,7 @@ def test_trivial(program_processor, lift_mode):
     )
 
     if validate:
-        assert np.allclose(out[:, :, 0], out_s)
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/iterator_tests/test_trivial.py:69]: undo, as `__array__` is now there.

File: tests/next_tests/integration_tests/multi_feature_tests/ffront_tests/test_laplacian.py [hunk 6]
```diff
@@ -69,8 +69,8 @@ def test_ffront_lap(cartesian_case):
         lap_program,
         in_field,
         out_field,
-        inout=out_field.array()[1:-1, 1:-1],
-        ref=lap_ref(np.asarray(in_field)),
+        inout=out_field.ndarray[1:-1, 1:-1],
```
Comment (havogt) [inline tests/next_tests/integration_tests/multi_feature_tests/ffront_tests/test_laplacian.py:72]: here we might want to express with named (relative?) ranges again

File: tests/next_tests/integration_tests/feature_tests/iterator_tests/test_builtins.py (hunk 1 above)
Comment (havogt) [reply tests/next_tests/integration_tests/feature_tests/iterator_tests/test_builtins.py:366]: np.asarray()

File: src/gt4py/next/embedded/nd_array_field.py [hunk 7]
```diff
@@ -133,14 +133,29 @@ def register_builtin_func(
...
+
+    @property
+    def __gt_origin__(self) -> tuple[int, ...]:
+        return tuple(-r.start for _, r in self._domain)
+
     @property
     def ndarray(self) -> core_defs.NDArrayObject:
         return self._ndarray
 
+    def __array__(self) -> np.ndarray:
```
Comment (havogt) [inline src/gt4py/next/embedded/nd_array_field.py:148]: check signature `dtype`?

File: src/gt4py/next/embedded/nd_array_field.py [hunk 8]
```diff
@@ -133,14 +133,29 @@ def register_builtin_func(
...
+        return np.asarray(self._ndarray)
+
     @property
     def value_type(self) -> type[core_defs.ScalarT]:
         return self._value_type
```
Comment (havogt) [inline src/gt4py/next/embedded/nd_array_field.py:153]:
```suggestion
    @property
    def value_type(self) -> type[core_defs.ScalarT]:
        return self.dtype.type
```
in `common.Field` or remove

File: src/gt4py/next/embedded/nd_array_field.py [hunk 9]
```diff
@@ -67,7 +67,7 @@ def _builtin_binary_op(a: _BaseNdArrayField, b: common.Field) -> common.Field:
                 return a.__class__.from_array(new_data, domain=domain_intersection)
             new_data = op(a.ndarray, xp.asarray(b.ndarray))
         else:
-            assert isinstance(b, core_defs.SCALAR_TYPES)
+            # assert isinstance(b, core_defs.SCALAR_TYPES) # TODO reenable this assert (if b is an array it should be wrapped into a field)
```
Comment (havogt) [inline src/gt4py/next/embedded/nd_array_field.py:70]: TODO: check if this is still needed and why

File: src/gt4py/next/embedded/nd_array_field.py [hunk 10]
```diff
@@ -174,12 +189,57 @@ def from_array(
...
+
+    restrict = (
+        __getitem__  # type:ignore[assignment] # TODO(havogt) I don't see the problem that mypy has
+    )
+
+    def __setitem__(
+        self,
+        index: common.FieldSlice,
+        value: common.Field | core_defs.NDArrayObject | core_defs.ScalarT,
+    ) -> None:
```
Comment (havogt) [inline src/gt4py/next/embedded/nd_array_field.py:214]: This requires some testing and refactoring (to not duplicate the logic of getitem that much). I'd prefer to do it in a separate PR.

File: src/gt4py/next/embedded/nd_array_field.py [hunk 11]
```diff
@@ -328,6 +374,17 @@ class JaxArrayField(_BaseNdArrayField):
     common.field.register(jnp.ndarray, JaxArrayField.from_array)
 
 
+def _tuplize_field_slice(v: common.FieldSlice) -> common.FieldSlice:
+    if (
```
Comment (havogt) [inline src/gt4py/next/embedded/nd_array_field.py:378]:
```suggestion
    """
    Wraps a single index/slice/range into a tuple.
    
    Note: the condition is complex as `NamedRange`, `NamedIndex` are implemented as `tuple`.
    """
    if (
```

File: src/gt4py/next/iterator/embedded.py [hunk 12]
```diff
@@ -878,60 +888,53 @@ def make_in_iterator(
...
-        return self.array()[indices]
+        return self._ndarrayfield.__gt_dims__
+
+    def _promote_tags_to_dims_and_sort(
+        self, named_indices: NamedFieldIndices
+    ) -> Mapping[common.Dimension, FieldIndex | SparsePositionEntry]:
+        return {d: named_indices[d.value] for d in self._ndarrayfield.__gt_dims__}
+
+    def _translate_named_indices(self, _named_indices: NamedFieldIndices) -> common.DomainSlice:
+        named_indices = self._promote_tags_to_dims_and_sort(_named_indices)
```
Comment (egparedes) [inline src/gt4py/next/iterator/embedded.py:907]: I would inline the function here (with the right type hint explicitly declared for named_indices) unless it results can be cached or something else is done in its body

File: src/gt4py/next/iterator/embedded.py [hunk 13]
```diff
@@ -878,60 +888,53 @@ def make_in_iterator(
...
     def field_setitem(self, named_indices: NamedFieldIndices, value: Any):
-        self.setter(get_ordered_indices(self._axes, named_indices), value)
+        if common.is_mutable_field(self._ndarrayfield):
+            self._ndarrayfield[self._translate_named_indices(named_indices)] = value
+        else:
+            raise RuntimeError("Assigment into a non-mutable Field.")
 
-    def __array__(self) -> np.ndarray:
-        return self.array()
+    def __array__(self) -> core_defs.NDArrayObject:
```
Comment (egparedes) [inline src/gt4py/next/iterator/embedded.py:932]: Incomplete signature (missing dtype argument)

File: src/gt4py/next/embedded/nd_array_field.py [hunk 14]
```diff
@@ -174,12 +189,57 @@ def from_array(
...
+
+    def __setitem__(
+        self,
+        index: common.FieldSlice,
+        value: common.Field | core_defs.NDArrayObject | core_defs.ScalarT,
+    ) -> None:
+        index = _tuplize_field_slice(index)
+        if common.is_field(value):
+            # TODO(havogt): in case of `is_field(value)` we should additionally check that `value.domain == self[slice].domain`
+            value = value.ndarray
```
Comment (egparedes) [inline src/gt4py/next/embedded/nd_array_field.py:218]: I would stick to the usual pattern of putting the body of the function in `restrict` and the defining `__getitem__` as `restrict`

**Review by egparedes — CHANGES_REQUESTED** [review]
I have some comments and questions, but overall looks good to me.

File: src/gt4py/_core/definitions.py [hunk 15]
```diff
@@ -330,13 +330,22 @@ def dtype(dtype_like: DTypeLike) -> DType:
...
+class StrLike(Protocol):
+    def __str__(self) -> str:
+        ...
+
+
 # -- Custom protocols  --
 class GTDimsInterface(Protocol):
-    __gt_dims__: Tuple[str, ...]
+    @property
+    def __gt_dims__(self) -> Tuple[str | StrLike, ...]:
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/_core/definitions.py:341]: this would introduce a circular dependency between the modules

File: src/gt4py/next/__init__.py [hunk 16]
```diff
@@ -25,6 +25,7 @@
 
 from . import common, ffront, iterator, program_processors, type_inference
 from .common import Dimension, DimensionKind, Field, GridType
+from .embedded import nd_array_field
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/__init__.py:28]: to register in the `common.field` dispatcher

File: src/gt4py/_core/definitions.py [hunk 17]
```diff
@@ -397,6 +406,9 @@ def dtype(self) -> Any:
     def __getitem__(self, item: SliceLike) -> NDArrayObject:
         ...
 
+    def __setitem__(self, item: SliceLike, value: Any) -> None:
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/_core/definitions.py:409]: Makes sense, removed. (Although jax arrays implement __setitem__ to raise an error.)

File: src/gt4py/next/common.py [hunk 18]
```diff
@@ -229,14 +252,24 @@ def __call__(self, func: fbuiltins.BuiltInFunction[_R, _P], /) -> Callable[_P, _
             ...
 
 
+class NextGTDimsInferface(Protocol):
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:255]: I had to read this 5 times to spot the typo.

File: src/gt4py/next/common.py [hunk 19]
```diff
@@ -229,14 +252,24 @@ def __call__(self, func: fbuiltins.BuiltInFunction[_R, _P], /) -> Callable[_P, _
...
     __gt_builtin_func__: ClassVar[GTBuiltInFuncDispatcher]
 
     @property
     def domain(self) -> Domain:
         ...
 
+    @property
+    def shape(self) -> tuple[int, ...]:  # TODO discuss this
+        return tuple(len(r) for r in self.domain.ranges)
+
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:272]: Doesn't make sense here, but maybe for NDArrayField? This is for use in test code outside of GT4Py programs. I'll move it for now.

File: src/gt4py/next/embedded/nd_array_field.py [hunk 20]
```diff
@@ -174,12 +189,57 @@ def from_array(
...
+
+        raise IndexError(f"Unsupported index type: {index}")
+
     __call__ = None  # type: ignore[assignment]  # TODO: remap
 
     __abs__ = _make_unary_array_field_intrinsic_func("abs", "abs")
 
     __neg__ = _make_unary_array_field_intrinsic_func("neg", "negative")
 
+    def __pos__(self):
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/embedded/nd_array_field.py:240]: I don't know, I'll change it.

File: src/gt4py/next/embedded/nd_array_field.py [hunk 21]
```diff
@@ -194,27 +254,13 @@ def remap(self: _BaseNdArrayField, connectivity) -> _BaseNdArrayField:
...
+    __and__ = __rand__ = _make_binary_array_field_intrinsic_func("bitwise_and", "bitwise_and")
+    __or__ = __ror__ = _make_binary_array_field_intrinsic_func("bitwise_or", "bitwise_or")
+    __xor__ = __rxor__ = _make_binary_array_field_intrinsic_func("bitwise_xor", "bitwise_xor")
 
-        raise IndexError(f"Unsupported index type: {index}")
-
-    restrict = (
-        __getitem__  # type:ignore[assignment] # TODO(havogt) I don't see the problem that mypy has
-    )
+    __invert__ = _make_unary_array_field_intrinsic_func("invert", "invert")
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/embedded/nd_array_field.py:263]: In field view we introduced these operators for boolean arrays (following numpy), but here I actually did the wrong thing for whatever reason.

File: src/gt4py/next/embedded/nd_array_field.py [hunk 22]
```diff
@@ -328,6 +374,17 @@ class JaxArrayField(_BaseNdArrayField):
     common.field.register(jnp.ndarray, JaxArrayField.from_array)
 
 
+def _tuplize_field_slice(v: common.FieldSlice) -> common.FieldSlice:
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/embedded/nd_array_field.py:377]: not sure. you mean `tupleize`, but then the `ei` gets collapsed at least sometimes (e.g. concrete -> concretize)

File: src/gt4py/next/iterator/embedded.py (hunk 13 above)
Comment (havogt) [reply src/gt4py/next/iterator/embedded.py:932]: Trying to delete it, not sure if it's actually used.

File: src/gt4py/_core/definitions.py [hunk 23]
```diff
@@ -330,13 +330,22 @@ def dtype(dtype_like: DTypeLike) -> DType:
     return dtype_like if isinstance(dtype_like, DType) else DType(np.dtype(dtype_like).type)
 
 
+class StrLike(Protocol):
+    def __str__(self) -> str:
+        ...
+
+
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/_core/definitions.py:345]: removed (and postponed how to unify with NextGTDimsInterface)

File: src/gt4py/_core/definitions.py [hunk 24]
```diff
@@ -251,6 +246,14 @@ def lanes(self) -> int:
     def subndim(self) -> int:
         return len(self.tensor_shape)
 
+    def __eq__(self, other: Any) -> bool:
+        # TODO: discuss (make concrete subclasses equal to instances of this with the same type)
```
Comment (havogt) [inline src/gt4py/_core/definitions.py:250]: TODO

File: src/gt4py/next/iterator/embedded.py [hunk 25]
```diff
@@ -1443,14 +1417,9 @@ def field_setitem(self, named_indices: NamedFieldIndices, value: Any):
...
-    assert can_be_tuple_field(field)
-
-    if is_tuple_of_field(field):
-        assert not isinstance(field, TupleField)
-        return TupleOfFields(tuple(_wrap_field(f) for f in field))
-
-    assert isinstance(field, TupleField)  # e.g. field of tuple is already TupleField
-    return field
+    assert is_tuple_of_field(field)
+    assert not isinstance(field, TupleField)
```
Comment (havogt) [inline src/gt4py/next/iterator/embedded.py:1415]: This removes support for structured dtypes (tests where already removed earlier in this PR).

File: src/gt4py/next/embedded/nd_array_field.py [hunk 26]
```diff
@@ -174,12 +189,57 @@ def from_array(
...
+    )
+
+    def __setitem__(
+        self,
+        index: common.FieldSlice,
+        value: common.Field | core_defs.NDArrayObject | core_defs.ScalarT,
+    ) -> None:
+        index = _tuplize_field_slice(index)
+        if common.is_field(value):
+            # TODO(havogt): in case of `is_field(value)` we should additionally check that `value.domain == self[slice].domain`
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/embedded/nd_array_field.py]: I would like to do a cleanup around the getitem/setitem related to the changes we discussed, therefore I don't want to do too much here. Maybe we should do that refactoring PR first.

File: src/gt4py/next/embedded/nd_array_field.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/next/embedded/nd_array_field.py:408]: You're right, `tuplize` makes more sense.

File: tests/next_tests/integration_tests/cases.py [hunk 27]
```diff
@@ -75,7 +74,7 @@
 C2E = gtx.FieldOffset("E2V", source=Edge, target=(Cell, C2EDim))
 
 ScalarValue: TypeAlias = np.int32 | np.int64 | np.float32 | np.float64 | np.generic
-FieldValue: TypeAlias = gtx.Field | embedded.LocatedFieldImpl
+FieldValue: TypeAlias = gtx.Field  # | embedded.LocatedFieldImpl
```
Comment (egparedes) [inline tests/next_tests/integration_tests/cases.py:77]: I would either add an explanation for the commented out code or delete the comment.

File: src/gt4py/next/iterator/embedded.py [hunk 28]
```diff
@@ -744,48 +733,51 @@ def _make_tuple(
...
-            )
-            col[0] = first
-            for i in column_range[1:]:
-                col[i] = tuple(
-                    _make_tuple(f, _single_vertical_idx(named_indices, column_axis, i))
-                    for f in field_or_tuple
-                )
-            return col
-        else:
+) -> Column | npt.DTypeLike | tuple[tuple | Column | npt.DTypeLike, ...] | Undefined:
```
Comment (egparedes) [inline src/gt4py/next/iterator/embedded.py:736]: Just a question: shouldn't  the overloads accepting LocatedField return an union including `Undefined` ?

File: src/gt4py/next/iterator/embedded.py [hunk 29]
```diff
@@ -744,48 +733,51 @@ def _make_tuple(
...
+                )
+            except embedded_exceptions.IndexOutOfBounds:
+                col.append(_UNDEFINED)
 
-def _axis_idx(axes: Sequence[common.Dimension], axis: Tag) -> Optional[int]:
-    for i, a in enumerate(axes):
-        if a.value == axis:
-            return i
-    return None
+        first = next((v for v in col if v != _UNDEFINED))
```
Comment (egparedes) [inline src/gt4py/next/iterator/embedded.py:773]: Is it guaranteed that there is always at least one `v != _UNDEFINED`  ?

**Review by egparedes — APPROVED** [review]
I have a couple of questions but they are not critical so I'm approving the PR anyway.

File: src/gt4py/next/iterator/embedded.py (hunk 28 above)
Comment (havogt) [reply src/gt4py/next/iterator/embedded.py:736]: yes

File: src/gt4py/next/iterator/embedded.py (hunk 29 above)
Comment (havogt) [reply src/gt4py/next/iterator/embedded.py:773]: It should not for a legal program, but I added a check now.

File: tests/next_tests/integration_tests/cases.py (hunk 27 above)
Comment (havogt) [reply tests/next_tests/integration_tests/cases.py:77]: deleted

File: src/gt4py/next/iterator/embedded.py (hunk 28 above)
Comment (havogt) [reply src/gt4py/next/iterator/embedded.py:736]: adding it introduced a type error that I didn't want to sort out now. added a TODO...

Comment (havogt) [reply src/gt4py/next/iterator/embedded.py:736]: found the problem..
